src/backend/auxil/range_alloc.rs

```diff
@@ -0,0 +1,169 @@
+        }
+        best_range_tuple.map(|range| {
+            if range.1.end - range.1.start == length {
+                self.empty_ranges.remove(range.0);
+            } else {
+                self.empty_ranges[range.0].start += length;
+            }
+            range.1.start..(range.1.start + length)
+        })
+    }
+
+    pub fn free_range(&mut self, range: Range<T>) -> bool {
+        assert!(range.start >= self.range.start && range.end <= self.range.end);
+        assert!(range.start <= range.end);
+        if self.empty_ranges.len() == 0 {
+            self.empty_ranges.push(range);
+            return true;
+        }
+        // Input is within range, but before any empty ranges
+        if self.empty_ranges.len() > 0 {
+            if self.empty_ranges[0].start == range.end {
+                self.empty_ranges[0].start = range.start;
+                return true;
+            } else if self.empty_ranges[0].start > range.end {
+                self.empty_ranges.insert(0, range);
+                return true;
+            }
+        }
+        // Input is within range, but after all empty ranges
+        if let Some(last) = self.empty_ranges.last().cloned() {
+            if last.end == range.start {
+                let index = self.empty_ranges.len() - 1;
+                self.empty_ranges[index].end = range.end;
+                return true;
+            } else if last.end < range.start {
+                self.empty_ranges.push(range);
+                return true;
+            }
+        }
+        for i in 0..self.empty_ranges.len() {
+            // Input is immediately to the left of an existing empty range.
+            if range.end == self.empty_ranges[i].start {
+                // Extend this range
+                self.empty_ranges[i].start = range.start;
+                // Merge this into an adjacent range to the left if necessary.
+                if i > 0 && self.empty_ranges[i - 1].end == self.empty_ranges[i].start {
+                    let r = self.empty_ranges.remove(i);
+                    self.empty_ranges[i - 1].end = r.end;
+                    
+                }
+                return true;
+            }
+
+            // Input is immediately to the right of an existing empty range.
+            if range.start == self.empty_ranges[i].end {
+                // Extend this range
+                self.empty_ranges[i].end = range.end;
+
+                // Merge this into an adjacent range to the right if necessary.
+                if i + 1 != self.empty_ranges.len() && self.empty_ranges[i + 1].start == self.empty_ranges[i].end {
+                    let r = self.empty_ranges.remove(i + 1);
+                    self.empty_ranges[i].end = r.end;
+                }
+                return true;
+            }
+
+            // Input is inbetween two empty ranges.
+            if i + 1 != self.empty_ranges.len() && range.start > self.empty_ranges[i].end && range.end < self.empty_ranges[i + 1].start {
```

Comment:
just check for `empty[i].start < start` based on the sorting constraint.
Also do this check on the beginning of the loop to just find the position where to insert the range.
Ideally it would look like this:
```rust
idx = self.empty_ranges.iter().positon(|r| r.start < range.start)
// check left merge
// check right merge
// insert

Reply:
I probably haven't explained my motivation properly behind this. The idea was to shrinking the loop to a minium, only searching for the insertion position. All the merge/insertion/removal operations would then only operate with the returned index.

(edit: my condition inside position(..) might be incorrect..)

Reply:
Uh yeah now that I look at that condition again I'm not sure it's possible to create correct code with the `position` function, so I'm going to undo this change.